teps/0017-shell-escaped-params.md

```diff
@@ -0,0 +1,142 @@
+---
+title: Provide Shell-Escaped Parameters
+authors:
+  - "@coryrc"
+creation-date: 2020-09-15
+last-updated: 2020-09-15
+status: proposed
+---
+
+# TEP-0017 Provide Shell-Escaped Parameters
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+It is difficult to safely use parameters in a `script:` scalar. This TEP proposes
+offering a shell-escaped version of every parameter.
+
+## Motivation
+
+- More concise than the current workarounds
+- Help users create more robust Tasks by default
+- More clearly expresses intent
+
+### Goals
+
+Provide a way to place parameters directly into the default `script:` without being
+interpreted by the shell.
+
+### Non-Goals
+
+1. Prevent purposeful interpretation of parameters when intended
+2. Work with shells which are not the default for the `script:`
+3. Be perfectly secure against arbitrary data
+
+## Requirements
+
+## Proposal
+
+In a Task or Pipeline, for each parameter `foo`, provide `$(params.foo.shell-escaped)`.
+
+This variable shall be escaped according to `printf '%q'` rules as used in bash.
+
+Support Bourne, Bourne-Again, Busybox, and Debian Almquist shells against
+injection, but don't purposefully prevent its usage in other scenarios.
```

Comment:
could you explain a bit more about what exactly this would look like, maybe with some examples?

i.e. given input X, the task would be provided with Y

Looking at the printf command I don't fully understand what this implies:

> Output the corresponding argument in a format that can be reused as shell input

Does this mean we need to know what shell we're dealing with and output differently according to that, or will the result be the same for every shell?

Reply:
The result will be the same for every shell.
I will put in an example.

Reply:
I added answers to all your questions.